registry/auth/basic/access.go

```diff
@@ -0,0 +1,92 @@
+// Package basic provides a simple authentication scheme that checks for the
+// user credential hash in an htpasswd formatted file in a configuration-determined
+// location.
+//
+// This authentication method MUST be used under TLS, as simple token-replay attack is possible.
+package basic
+
+import (
+	"errors"
+	"fmt"
+	"net/http"
+
+	ctxu "github.com/docker/distribution/context"
+	"github.com/docker/distribution/registry/auth"
+	"golang.org/x/net/context"
+)
+
+type accessController struct {
+	realm    string
+	htpasswd *htpasswd
+}
+
+type challenge struct {
+	realm string
+	err   error
+}
+
+var _ auth.AccessController = &accessController{}
+var (
+	// ErrPasswordRequired Returned when no auth token is given.
+	ErrPasswordRequired = errors.New("authorization credential required")
+	// ErrInvalidCredential is returned when the auth token does not authenticate correctly.
+	ErrInvalidCredential = errors.New("invalid authorization credential")
+)
+
+func newAccessController(options map[string]interface{}) (auth.AccessController, error) {
+	realm, present := options["realm"]
+	if _, ok := realm.(string); !present || !ok {
+		return nil, fmt.Errorf(`"realm" must be set for basic access controller`)
+	}
+
+	path, present := options["path"]
+	if _, ok := path.(string); !present || !ok {
+		return nil, fmt.Errorf(`"path" must be set for basic access controller`)
+	}
+
+	return &accessController{realm: realm.(string), htpasswd: newHTPasswd(path.(string))}, nil
+}
+
+func (ac *accessController) Authorized(ctx context.Context, accessRecords ...auth.Access) (context.Context, error) {
+	req, err := ctxu.GetRequest(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	authHeader := req.Header.Get("Authorization")
+	if authHeader == "" {
+		challenge := challenge{
+			realm: ac.realm,
+		}
+		return nil, &challenge
+	}
+
+	user, pass, ok := req.BasicAuth()
+	if !ok {
+		return nil, errors.New("Invalid Authorization header")
+	}
+
+	if res, _ := ac.htpasswd.AuthenticateUser(user, pass); !res {
```

Comment:
I think an error should be checked here and be logged if it is not ```nil``` or not ```EOF```. For example, it would be hard to debug a problem, when we have no rights to read the password file.